### PR TITLE
Pluggable resource filter

### PR DESCRIPTION
**Pull request by yumaoka**
Redesigned resource filter interface and introduced resource filter service that allows users to provide their own custom filter implementations.

### CODE REVIEW

**Comment by yumaoka** [comment]
This is still WIP. I'm going to add

- test cases for example custom filter.
- example maven/ant project utilizing the custom filter.

File: gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/BundleSet.java [hunk 1]
```diff
@@ -33,7 +31,7 @@
  */
 public class BundleSet {
 
-    private ResourceType type = ResourceType.JAVA;
+    private String type = "JAVA";
```
Comment (srl295) [inline gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/BundleSet.java]: change constant to `DefaultResourceFilterProvider.Filter.JAVA.name()`?

File: gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/GPBaseTask.java [hunk 2]
```diff
@@ -234,14 +233,14 @@ public BundleSet createBundleSet() {
...
-    private String pathToBundleId(ResourceType type, String path) {
+    private String pathToBundleId(String type, String path) {
         File f = new File(path);
         File parent = f.getParentFile();
         String pkgName = parent == null ? "" :
             parent.getPath().replace(File.separatorChar, '.');
 
         String fileName = f.getName().replaceAll(" ", "_");
-        if (type == ResourceType.JAVA) {
+        if (type.equals("JAVA")) {
```
Comment (srl295) [inline gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/GPBaseTask.java]: change constant to `DefaultResourceFilterProvider.Filter.JAVA.name()`?

File: gp-maven-plugin/src/main/java/com/ibm/g11n/pipeline/maven/BundleSet.java [hunk 3]
```diff
@@ -35,7 +33,7 @@
     private FileSet sourceFiles;
 
     @Parameter(defaultValue = "JAVA")
```
Comment (srl295) [inline gp-maven-plugin/src/main/java/com/ibm/g11n/pipeline/maven/BundleSet.java]: same comment… change constant to DefaultResourceFilterProvider.Filter.JAVA.name()?

File: gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/BundleSet.java (hunk 1 above)
Comment (srl295) [reply gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/BundleSet.java]: also , is this really an `id` and not a type?

File: gp-res-filter/src/test/java/com/ibm/g11n/pipeline/resfilter/impl/GlobalizeJsResourceTest.java [hunk 4]
```diff
@@ -57,67 +63,61 @@
...
 
         try (OutputStream os = new FileOutputStream(tempFile)) {
-            res.write(os, "de", WRITE_BUNDLE);
+            res.write(os, WRITE_BUNDLE, new FilterOptions(Locale.GERMAN));
             os.flush();
             assertTrue(ResourceTestUtil.compareFiles(EXPECTED_WRITE_FILE, tempFile));
         }
     }
 
+ // TODO: Not ready yet
```
Comment (srl295) [inline gp-res-filter/src/test/java/com/ibm/g11n/pipeline/resfilter/impl/GlobalizeJsResourceTest.java]: TODO - does this need a new issue?

File: gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/GPBaseTask.java (hunk 2 above)
Comment (srl295) [reply gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/GPBaseTask.java]: is this also an 'id' where type is single or multi

File: gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/BundleSet.java (hunk 1 above)
Comment (yumaoka) [reply gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/BundleSet.java]: Previously we used ResourceType enum. But enum does not fit well with dynamic service registration. I agree we have better way to define this.

`id` vs. `type`. I think it's resource type ID. We already use the term type in Maven/Ant interface, I cannot change the term in Maven/Ant easily.

Comment (srl295) [reply gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/BundleSet.java]: I'm looking at https://github.com/IBM-Cloud/gp-java-tools/pull/93/files#diff-4199ca2c76f4388d7b95620e3e72b839R44 where `CSV-MULTI` is the ID, and `Type.MULTI` is the type.  I think it was confusing where in the code 'type' refers to the filter ID

File: gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/GPBaseTask.java (hunk 2 above)
Comment (yumaoka) [reply gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/GPBaseTask.java]: Same as above.

I was thinking about both single bundle and multi bundle filter implementation can be provided through a single resource filter ID. But I decided not to do so.

So, for a given filter ID "foo", it is either a single bundle filter or a multiple bundle filter, but cannot do both.

File: gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/BundleSet.java (hunk 1 above)
Comment (srl295) [reply gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/BundleSet.java]: sorry, this class is within ant, so maybe not relevant.

Comment (srl295) [reply gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/BundleSet.java]: sorry, this is in the ant class, so probably OK

File: gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/GPBaseTask.java (hunk 2 above)
Comment (srl295) [reply gp-ant-task/src/main/java/com/ibm/g11n/pipeline/ant/GPBaseTask.java]: also never mind, this is not the filter id but bundle type within ant.  

Ignore this comment ^

File: gp-maven-plugin/src/main/java/com/ibm/g11n/pipeline/maven/BundleSet.java (hunk 3 above)
Comment (yumaoka) [reply gp-maven-plugin/src/main/java/com/ibm/g11n/pipeline/maven/BundleSet.java]: I'll find out. I agree that hardcoding string here is not ideal.

File: gp-res-filter/src/test/java/com/ibm/g11n/pipeline/resfilter/impl/GlobalizeJsResourceTest.java (hunk 4 above)
Comment (yumaoka) [reply gp-res-filter/src/test/java/com/ibm/g11n/pipeline/resfilter/impl/GlobalizeJsResourceTest.java]: "merge" is optional operation, and if not implemented, it's same as "write". GlobalizeJsResource does not have implementation. But we "may" implement it later time.  I just added a comment - to be addressed later.

**Pull request merged**